File: utils/mqtt/MqttClient.java
import org.eclipse.paho.client.mqttv3.*;
import org.eclipse.paho.client.mqttv3.persist.MemoryPersistence;
import com.google.gson.Gson;

/**
 * MQTT 客户端 - Java
 * 
 * Maven 依赖:
 * <dependency>
 *     <groupId>org.eclipse.paho</groupId>
 *     <artifactId>org.eclipse.paho.client.mqttv3</artifactId>
 *     <version>1.2.5</version>
 * </dependency>
 * 
 * 编译运行:
 * mvn compile exec:java -Dexec.mainClass="MqttClient"
 */

public class MqttClient {
    
    // ============== 配置区域 ==============
    private static final String BROKER = "tcp://localhost:1883";
    private static final String CLIENT_ID = "mqtt_java_client_" + System.currentTimeMillis();
    private static final String USERNAME = "";  // 用户名（如果需要）
    private static final String PASSWORD = "";  // 密码（如果需要）
    private static final String TOPIC = "test/topic";
    private static final int QOS = 1;
    // =====================================
    
    private IMqttAsyncClient client;
    private Gson gson = new Gson();
    private MessageCallback messageCallback;
    
    /**
     * 消息回调接口
     */
    public interface MessageCallback {
        void onMessage(String topic, String message);
    }
    
    /**
     * 连接到 MQTT 代理
     */
    public MqttClient connect() {
        try {
            MemoryPersistence persistence = new MemoryPersistence();
            client = new MqttAsyncClient(BROKER, CLIENT_ID, persistence);
            
            // 设置回调
            client.setCallback(new MqttCallback() {
                @Override
                public void connectionLost(Throwable cause) {
                    System.out.println("[已断开] 连接丢失: " + cause.getMessage());
                }
                
                @Override
                public void messageArrived(String topic, MqttMessage message) throws Exception {
                    String payload = new String(message.getPayload());
                    System.out.println("[收到消息] 主题: " + topic + ", 消息: " + payload);
                    
                    if (messageCallback != null) {
                        messageCallback.onMessage(topic, payload);
                    }
                }
                
                @Override
                public void deliveryComplete(IMqttDeliveryToken token) {
                    // 消息发送完成
                }
            });
            
            // 连接选项
            MqttConnectOptions options = new MqttConnectOptions();
            options.setCleanSession(true);
            options.setConnectionTimeout(10);
            options.setKeepAliveInterval(60);
            
            if (!USERNAME.isEmpty()) {
                options.setUserName(USERNAME);
            }
            if (!PASSWORD.isEmpty()) {
                options.setPassword(PASSWORD.toCharArray());
            }
            
            // 连接
            client.connect(options).waitForCompletion();
            System.out.println("[已连接] 连接到代理: " + BROKER);
            
        } catch (MqttException e) {
            System.err.println("[连接错误] " + e.getMessage());
        }
        
        return this;
    }
    
    /**
     * 发布消息
     * @param topic 主题
     * @param message 消息内容
     * @param qos 服务质量等级 (0, 1, 2)
     */
    public void publish(String topic, Object message, int qos) {
        if (client == null || !client.isConnected()) {
            System.err.println("[错误] 客户端未连接");
            return;
        }
        
        try {
            String payload;
            if (message instanceof String) {
                payload = (String) message;
            } else {
                payload = gson.toJson(message);
            }
            
            MqttMessage mqttMessage = new MqttMessage(payload.getBytes());
            mqttMessage.setQos(qos);
            
            client.publish(topic, mqttMessage).waitForCompletion();
            System.out.println("[已发布] 主题: " + topic + ", 消息: " + payload);
            
        } catch (MqttException e) {
            System.err.println("[发布失败] " + e.getMessage());
        }
    }
    
    /**
     * 发布消息（使用默认 QOS）
     */
    public void publish(String topic, Object message) {
        publish(topic, message, QOS);
    }
    
    /**
     * 订阅主题
     * @param topics 主题数组
     * @param callback 消息回调
     */
    public void subscribe(String[] topics, MessageCallback callback) {
        if (client == null || !client.isConnected()) {
            System.err.println("[错误] 客户端未连接");
            return;
        }
        
        this.messageCallback = callback;
        
        try {
            int[] qos = new int[topics.length];
            for (int i = 0; i < topics.length; i++) {
                qos[i] = QOS;
            }
            
            client.subscribe(topics, qos).waitForCompletion();
            System.out.println("[已订阅] 主题: " + String.join(", ", topics));
            
        } catch (MqttException e) {
            System.err.println("[订阅失败] " + e.getMessage());
        }
    }
    
    /**
     * 订阅单个主题
     */
    public void subscribe(String topic, MessageCallback callback) {
        subscribe(new String[]{topic}, callback);
    }
    
    /**
     * 取消订阅
     * @param topics 主题数组
     */
    public void unsubscribe(String[] topics) {
        if (client == null || !client.isConnected()) {
            System.err.println("[错误] 客户端未连接");
            return;
        }
        
        try {
            client.unsubscribe(topics).waitForCompletion();
            System.out.println("[已取消订阅] 主题: " + String.join(", ", topics));
            
        } catch (MqttException e) {
            System.err.println("[取消订阅失败] " + e.getMessage());
        }
    }
    
    /**
     * 断开连接
     */
    public void disconnect() {
        if (client != null && client.isConnected()) {
            try {
                client.disconnect().waitForCompletion();
                client.close();
                System.out.println("[已断开] 客户端已断开连接");
            } catch (MqttException e) {
                System.err.println("[断开错误] " + e.getMessage());
            }
        }
    }
    
    /**
     * 检查是否已连接
     */
    public boolean isConnected() {
        return client != null && client.isConnected();
    }
    
    // ============== 使用示例 ==============
    public static void main(String[] args) throws InterruptedException {
        MqttClient mqttClient = new MqttClient();
        
        // 连接
        mqttClient.connect();
        
        // 等待连接建立
        Thread.sleep(1000);
        
        // 订阅主题
        mqttClient.subscribe(TOPIC, (topic, message) -> {
            System.out.println("回调收到: " + topic + " -> " + message);
        });
        
        // 发布消息
        mqttClient.publish(TOPIC, "Hello MQTT from Java!");
        mqttClient.publish(TOPIC, new TestData("json", "test"));
        
        // 保持运行
        Thread.sleep(10000);
        
        // 断开连接
        mqttClient.disconnect();
    }
    
    // 测试数据类
    static class TestData {
        String type;
        String data;
        
        TestData(String type, String data) {
            this.type = type;
            this.data = data;
        }
    }
}

File: utils/mqtt/mqtt_client.go
// MQTT 客户端 - Go
//
// 依赖安装: go get github.com/eclipse/paho.mqtt.golang
// 运行: go run mqtt_client.go

package main

import (
	"encoding/json"
	"fmt"
	"os"
	"os/signal"
	"syscall"
	"time"

	mqtt "github.com/eclipse/paho.mqtt.golang"
)

// ============== 配置区域 ==============
var config = MqttConfig{
	Broker:   "tcp://localhost:1883",
	ClientID: fmt.Sprintf("mqtt_go_client_%d", time.Now().Unix()),
	Username: "", // 用户名（如果需要）
	Password: "", // 密码（如果需要）
	Topic:    "test/topic",
	Qos:      byte(1),
}

// MqttConfig MQTT 配置结构
type MqttConfig struct {
	Broker   string
	ClientID string
	Username string
	Password string
	Topic    string
	Qos      byte
}

// =====================================

// MqttClient MQTT 客户端
type MqttClient struct {
	client mqtt.Client
	config MqttConfig
}

// MessageHandler 消息处理函数类型
type MessageHandler func(topic string, message string)

// NewMqttClient 创建新的 MQTT 客户端
func NewMqttClient(config MqttConfig) *MqttClient {
	return &MqttClient{
		config: config,
	}
}

// Connect 连接到 MQTT 代理
func (m *MqttClient) Connect() *MqttClient {
	opts := mqtt.NewClientOptions()
	opts.AddBroker(m.config.Broker)
	opts.SetClientID(m.config.ClientID)
	opts.SetCleanSession(true)
	opts.SetAutoReconnect(true)

	if m.config.Username != "" {
		opts.SetUsername(m.config.Username)
	}
	if m.config.Password != "" {
		opts.SetPassword(m.config.Password)
	}

	// 设置连接回调
	opts.OnConnect = func(c mqtt.Client) {
		fmt.Printf("[已连接] 连接到代理: %s\n", m.config.Broker)
	}

	opts.OnConnectionLost = func(c mqtt.Client, err error) {
		fmt.Printf("[已断开] 连接丢失: %v\n", err)
	}

	m.client = mqtt.NewClient(opts)

	if token := m.client.Connect(); token.Wait() && token.Error() != nil {
		fmt.Printf("[连接错误] %v\n", token.Error())
	}

	return m
}

// Publish 发布消息
func (m *MqttClient) Publish(topic string, message interface{}, qos ...byte) {
	if !m.client.IsConnected() {
		fmt.Println("[错误] 客户端未连接")
		return
	}

	q := m.config.Qos
	if len(qos) > 0 {
		q = qos[0]
	}

	var payload string
	switch v := message.(type) {
	case string:
		payload = v
	default:
		bytes, err := json.Marshal(v)
		if err != nil {
			fmt.Printf("[序列化错误] %v\n", err)
			return
		}
		payload = string(bytes)
	}

	token := m.client.Publish(topic, q, false, payload)
	token.Wait()

	if token.Error() != nil {
		fmt.Printf("[发布失败] %v\n", token.Error())
	} else {
		fmt.Printf("[已发布] 主题: %s, 消息: %s\n", topic, payload)
	}
}

// Subscribe 订阅主题
func (m *MqttClient) Subscribe(topics []string, handler MessageHandler) {
	if !m.client.IsConnected() {
		fmt.Println("[错误] 客户端未连接")
		return
	}

	for _, topic := range topics {
		token := m.client.Subscribe(topic, m.config.Qos, func(c mqtt.Client, msg mqtt.Message) {
			payload := string(msg.Payload())
			fmt.Printf("[收到消息] 主题: %s, 消息: %s\n", msg.Topic(), payload)
			if handler != nil {
				handler(msg.Topic(), payload)
			}
		})
		token.Wait()

		if token.Error() != nil {
			fmt.Printf("[订阅失败] %v\n", token.Error())
		}
	}

	fmt.Printf("[已订阅] 主题: %v\n", topics)
}

// SubscribeSingle 订阅单个主题
func (m *MqttClient) SubscribeSingle(topic string, handler MessageHandler) {
	m.Subscribe([]string{topic}, handler)
}

// Unsubscribe 取消订阅
func (m *MqttClient) Unsubscribe(topics []string) {
	if !m.client.IsConnected() {
		fmt.Println("[错误] 客户端未连接")
		return
	}

	token := m.client.Unsubscribe(topics...)
	token.Wait()

	if token.Error() != nil {
		fmt.Printf("[取消订阅失败] %v\n", token.Error())
	} else {
		fmt.Printf("[已取消订阅] 主题: %v\n", topics)
	}
}

// Disconnect 断开连接
func (m *MqttClient) Disconnect() {
	if m.client.IsConnected() {
		m.client.Disconnect(250)
		fmt.Println("[已断开] 客户端已断开连接")
	}
}

// IsConnected 检查是否已连接
func (m *MqttClient) IsConnected() bool {
	return m.client.IsConnected()
}

// ============== 使用示例 ==============
func main() {
	mqttClient := NewMqttClient(config)

	// 连接
	mqttClient.Connect()

	// 等待连接建立
	time.Sleep(1 * time.Second)

	// 订阅主题
	mqttClient.SubscribeSingle(config.Topic, func(topic string, message string) {
		fmt.Printf("回调收到: %s -> %s\n", topic, message)
	})

	// 发布消息
	mqttClient.Publish(config.Topic, "Hello MQTT from Go!")
	mqttClient.Publish(config.Topic, map[string]string{
		"type": "json",
		"data": "test",
	})

	// 等待信号退出
	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)

	// 10秒后自动断开
	go func() {
		time.Sleep(10 * time.Second)
		mqttClient.Disconnect()
		sigChan <- syscall.SIGTERM
	}()

	<-sigChan
	fmt.Println("程序退出")
}